src/krb5/_principal.pyx

```diff
@@ -186,3 +392,47 @@ def unparse_name_flags(
         return <bytes>name
     finally:
         krb5_free_unparsed_name_generic(context.raw, name)
+
+
+# Note: build_principal() does not actually call krb5_build_principal() because
+# this would require passing vararg parameters and because
+# krb5_build_principal() cannot handle NUL bytes in the strings.
+# Instead, the principal is built manually, and then duplicated with
+# copy_principal().
+def build_principal(
+    Context context not None,
+    const unsigned char[:] realm not None,
+    components: typing.Iterable[bytes],
+) -> Principal:
+    component_list = list(components)
+    component_count = len(component_list)
+
+    # Guess the component count, similar to the MIT implementation starting with 1.12
+    if component_count == 2 and component_list[0] == b'krbtgt':
+        inferred_principal_type = 2  # KRB5_NT_SRV_INST
```

Comment:
I'm using the enum values now. (Same for the tests.)

I've hardcoded the enum values instead of using the `cdef extern from ...` block as these are on-the-wire values which should be the same for all implementations.